### PR TITLE
Documentation for Database Continuous Integration

### PR DESCRIPTION
**Pull request by garrettblehm**
Let me know if there are any typos or anything that needs to be changed.

### CODE REVIEW

File: codingstandards/continuousintegration.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Database Continuous Integration
+## Overview
+Database Continuous Integration (CI) is a method in which changes to the database are recorded incrementally in order to provide a continuously updated database. This form of integration allows for developers to have local copies of a database and to update their local database incrementally.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: I would add a comment about how this enables automated testing of data access code and repositories.

File: codingstandards/continuousintegration.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Database Continuous Integration
+## Overview
+Database Continuous Integration (CI) is a method in which changes to the database are recorded incrementally in order to provide a continuously updated database. This form of integration allows for developers to have local copies of a database and to update their local database incrementally.  
+
+## Install vs. Update
+For each database using the CI design there are two levels of versioning that need to be changed. They are `install` and `update`.
+
+### Install
+The `install` changes are all the changes needed to create the database from scratch. This will include all table declorations and stored procedures. The `install` changes will be applied whenever a devoloper wants to create a database. This reqires the `install` directory to have **the current state of the database**. All changes to the database will modify the `install` directory.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: **declorations** is misspelled.  I should be **declarations**.

File: codingstandards/continuousintegration.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+### Update
+The `update` changes are all the changes needed to get the database from point **A** to point **B**. The `update` changes are versioned with a unique modifier so that the database knows what updates it has completed and which ones it needs to run. The PROD database will run the `update` changes whenever a new change is released.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: How about something like "...get existing databases from point **A** to point **B**."

Also, the last sentence, be more clear that it isn't just PROD: "An existing environment, such as TEST or PROD, will have the `update` changes applied when the new version is released."

File: codingstandards/continuousintegration.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+### 1. Modify Install directory
+
+Modifiy the `*ProjectName*/Database/Install` directory to reflect the current state of the database. The `Install` directory has multiple subdirectories for the different components of the database. 
+
+![Install folder](/images/LegacyWebInstallFolder.png)
+
+* The `Functions` directory stores functions.
+* The `Routines` directory stores Stored Procedures.
+* The `Security` directory stores Schemas.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: Schemas, roles, etc.

File: codingstandards/continuousintegration.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+* The `Tables` directory stores Table Declorations.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: declarations

File: codingstandards/continuousintegration.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+## Performing a Change
+When making a change to a database using the CI design follow these steps.
+
+### 1. Modify Install directory
+
+Modifiy the `*ProjectName*/Database/Install` directory to reflect the current state of the database. The `Install` directory has multiple subdirectories for the different components of the database. 
+
+![Install folder](/images/LegacyWebInstallFolder.png)
+
+* The `Functions` directory stores functions.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: We usually put functions under procedures, perhaps in a subfolder.  If this isn't what has been practiced, though, it's not a biggie.

File: codingstandards/continuousintegration.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+Inside the `Install` directory is a C# file that controls what scripts are loaded on a database install. 
+
+![Migration file](/images/LegacyWebMigrationFile.png)
+
+A script is loaded with the command `Execute.EmbeddedScript("Database.Install.Tables.dbo.Customer.sql");`. This command takes as a parameter the path of the script. So the `Customer.sql` script is located at `*ProjectName*/Database/Install/Tables/dbo.Customer.sql`. 
+
+**When creating a new file or changing a filename make sure to add the script to the C# migration file.** When changing a file verify that it is in the C# migration file. 
+
+> Note: All scripts should be rerunnable. i.e. if a create table script is ran on a database that already has that table, that command will be ignored.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: I would include a code example of the migration for a fresh install.

File: codingstandards/continuousintegration.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+#### How to create migration C# file
+
+All C# migration files will import [FluentMigrator](https://github.com/fluentmigrator/fluentmigrator). They will also include a *Migration* tag that is formated according to the date and time of the change. This formatting ensures that migrations tags are unique. 
+
+![Update migration](/images/LegacyWebUpdateMigration.png)
+
+> Note: \[`Migration(201804091501)`\] would correspond to April 9, 2018 at 3:01 PM.
+
+The migration files will then reference the scripts that need to be run to execute the updates to the database.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: Include a code example of an update migration.

File: codingstandards/continuousintegration.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+* The `Tables` directory stores Table Declorations.
+* The `Types` directory stores Type Descriptions.
+
+Inside the `Install` directory is a C# file that controls what scripts are loaded on a database install. 
+
+![Migration file](/images/LegacyWebMigrationFile.png)
+
+A script is loaded with the command `Execute.EmbeddedScript("Database.Install.Tables.dbo.Customer.sql");`. This command takes as a parameter the path of the script. So the `Customer.sql` script is located at `*ProjectName*/Database/Install/Tables/dbo.Customer.sql`. 
+
+**When creating a new file or changing a filename make sure to add the script to the C# migration file.** When changing a file verify that it is in the C# migration file.
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md]: Also point out that the Build Action of the file should be "Embedded Resource".

File: codingstandards/continuousintegration.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+> Note: All scripts should be rerunnable. i.e. if a create table script is ran on a database that already has that table, that command will be ignored.
+
+### 2. Modify Update directory
+
+The `update` directory keeps track of changes that will advance the database from the previous state to the current state. Each update will be stored in it's own subdirectory. 
+
+![Update folder](/images/LegacyWebUpdateFolder.png)
+
+**Table changes will include SQL scripts** in the update subdirectories while **all other changes will only have the C# migration file**. 
+
```
Comment (johntkeller) [inline codingstandards/continuousintegration.md:96]: You can point out examples in a code sample mentioned below.

Comment (johntkeller) [inline codingstandards/continuousintegration.md:96]: Also point out that the Build Action of new SQL scripts should be "Embedded Resource".

**Review by johntkeller — CHANGES_REQUESTED** [review]
Thank you Garrett for putting this together.  Mostly minor comments, with the most significant being the addition of code examples.  I think that would go along way for pointing out the differences between install and update migrations.

File: codingstandards/continuousintegration.md (hunk 6 above)
Comment (garrettblehm) [reply codingstandards/continuousintegration.md]: I just looked at the LegacyWeb code in the packages. But I'll put the functions under procedures for the documentation!

File: codingstandards/continuousintegration.md (hunk 10 above)
Comment (garrettblehm) [reply codingstandards/continuousintegration.md:96]: Done.

Comment (garrettblehm) [reply codingstandards/continuousintegration.md:96]: Done.

**Review by johntkeller — APPROVED** [review]
Yep, I think this is a good first step.  If anyone finds need for clarification down the road, we can update as needed.  Thank you @GarrettB71 !

**Pull request merged**